### PR TITLE
feat: add atlas pull for the communications app FC-0012

### PR DESCRIPTION
**Pull request by OmarIthawi**
Adds `atlas pull` to the Communications MFE. Support for other MFEs will come.

## Previously open issues

Here's a couple issues that needs to be tackled:

 - Issue: How to enable the `atlas_pull` conditionally? 
   * Old solution: `CORE_MFE_APPS` feature.
   * Updated solution: Hard-coded `if app_name == communications`
 - Issue: How to install `openedx-atlas` and `platform-frontend` which are required for `atlas pull`?
   * Solution: In the package.json of the communications MFE.
 - Issue: Should we use `make pull translations`? or hand-craft an `atlas pull` command?
   * For discovery Regis recommended to avoid using make and writing an atlas command instead.
   * For MFEs, it's better to use the Makefile step to allow each frontend to control its own plugins.
 - Issue: After changing directory to `cd src/i18n/messages` the `node_modules/.bin` is no longer available.
   * Suggested solution: Use `ENV PATH /openedx/app/node_modules/.bin:${PATH}` in `base` instead of the relative PATH.
 - Issue: `atlas pull` is only available for master because https://github.com/openedx/frontend-app-communications/pull/124 is a post-Palm pull request. 
   * Patch has been done done on `palm.master`: https://github.com/openedx/frontend-app-communications/pull/156

## TODO before merge
 - [x] Test the translations and post a screenshot
    <details><summary>Screenshot: Je parle français! </summary>
    
    </details> 
 - [x] Backport the `atlas pull` to the Palm: https://github.com/openedx/frontend-app-communications/pull/156
- [x] Wait for Quince support and merge into Quince: https://github.com/overhangio/tutor-mfe/pull/156
 - [x] Verify that `atlas pull` and `intl-import.js` are both working as expected:
<details><summary>Debug atlas pull and the generated src/i18n/index.js</summary>

```
RUN ls -R /openedx/app/src/i18n/; cat /openedx/app/src/i18n/index.js; exit 1         0.4s
------                                                                                                                        
 > [communications-i18n 4/8] RUN ls -R /openedx/app/src/i18n/; cat /openedx/app/src/i18n/index.js; exit 1:                    
#0 0.385 /openedx/app/src/i18n/:                                                                                              
#0 0.385 index.js                                                                                                             
#0 0.385 messages                                                                                                             
#0 0.385                                                                                                                      
#0 0.385 /openedx/app/src/i18n/messages:
#0 0.385 frontend-app-communications
#0 0.385 frontend-component-footer
#0 0.385 frontend-component-header
#0 0.385 paragon
#0 0.385 
#0 0.385 /openedx/app/src/i18n/messages/frontend-app-communications:
#0 0.385 fr_CA.json
#0 0.385 index.js
#0 0.385 
#0 0.385 /openedx/app/src/i18n/messages/frontend-component-footer:
#0 0.385 ar.json
#0 0.385 de.json
#0 0.385 fr_CA.json
#0 0.385 index.js
#0 0.385 
#0 0.385 /openedx/app/src/i18n/messages/frontend-component-header:
#0 0.385 ar.json
#0 0.385 de.json
#0 0.385 fr_CA.json
#0 0.385 index.js
#0 0.385 
#0 0.385 /openedx/app/src/i18n/messages/paragon:
#0 0.385 ar.json
#0 0.385 de.json
#0 0.385 fr_CA.json
#0 0.385 index.js
#0 0.385 // This file is generated by the openedx/frontend-platform's "intl-import.js" script.
#0 0.385 //
#0 0.385 // Refer to the i18n documents in https://docs.openedx.org/en/latest/developers/references/i18n.html to update
#0 0.385 // the file and use the Micro-frontend i18n pattern in new repositories.
#0 0.385 //
#0 0.385 
#0 0.385 import messagesFromFrontendComponentHeader from './messages/frontend-component-header';
#0 0.385 import messagesFromFrontendComponentFooter from './messages/frontend-component-footer';
#0 0.385 import messagesFromParagon from './messages/paragon';
#0 0.385 import messagesFromFrontendAppCommunications from './messages/frontend-app-communications';
#0 0.385 
#0 0.385 export default [
#0 0.385   messagesFromFrontendComponentHeader,
#0 0.385   messagesFromFrontendComponentFooter,
#0 0.385   messagesFromParagon,
#0 0.385   messagesFromFrontendAppCommunications,
#0 0.385 ];
------
```

</details> 

## Nightly vs Master

All `atlas` integration is going to target `nightly` branches so it's included in Quince but not Palm.

The exception is for the Communications MFE because it has not translations and needs `atlas` in Palm, therefore it's being used on `master`.

References
----------
This contribution is part of the [FC-0012 project](https://openedx.atlassian.net/l/cp/XGS0iCcQ) which is sparked by the [Translation Infrastructure update OEP-58](https://open-edx-proposals.readthedocs.io/en/latest/architectural-decisions/oep-0058-arch-translations-management.html#specification).

### CODE REVIEW

**Comment by OmarIthawi** [comment]
@arbrandes I've spent few hours trying to get this working on Palm but couldn't because MFEs got their support for Atlas _after_ Palm.

I've also tried to work on nightly branches. It was successful, but then the Communications MFE shows a blank screen even without any changes to Tutor:

 - http://apps.local.overhang.io:1984/communications/courses/course-v1:s+s+s/bulk_email
 - image

File: tutormfe/plugin.py
Comment (OmarIthawi) [reply tutormfe/plugin.py:47]: yikes, that slipped through. Sorry.

**Review by OmarIthawi — COMMENTED** [review]
I'm adding notes to myself from my call with Adolfo.

**Comment by OmarIthawi** [comment]
@regisb @arbrandes I've tested this pull request which requires the following plugin:

```py
from tutormfe.hooks import MFE_APPS

@MFE_APPS.add()
def _use_comm_mfe_zeitlabs_fork(mfes):
    mfes["communications"] = {
        "repository": "https://github.com/Zeit-Labs/frontend-app-communications",
        "refs": "https://api.github.com/repos/Zeit-Labs/frontend-app-communications/git/refs/heads",
        "version": "atlas-palm",
        "port": 1984,
    }
    return mfes
```

So it pulls the right repository until the `palm.master` pull request is merged. 

Please review it and let me know what do you think.

**Comment by OmarIthawi** [comment]
@regisb @arbrandes would you mind giving this pull request one more look? Thanks!!

File: tutormfe/templates/mfe/build/mfe/Dockerfile
Comment (regisb) [inline tutormfe/templates/mfe/build/mfe/Dockerfile:75]: Are you familiar with the i18-merge.js script from line 53? It's a script that allows anyone to override translation strings. If I understand correctly, running `pull_translations` here would override this behaviour, thus preventing users from adding their custom translations. Can you confirm? Can we run Atlas in a way that is compatible with this existing feature?

Also, is atlas supported in Palm? If not, this PR should be opened on top of nightly.

Comment (OmarIthawi) [reply tutormfe/templates/mfe/build/mfe/Dockerfile:75]: > Are you familiar with the i18-merge.js script from line 53? It's a script that allows anyone to override translation strings. If I understand correctly, running pull_translations here would override this behaviour, thus preventing users from adding their custom translations. Can you confirm? Can we run Atlas in a way that is compatible with this existing feature?

I'm familiar with the script and I should have been clearer in the pull request description that we're going to intentionally not support `i18n-merge.js` and `atlas pull` at the same time in this iteration. This will affect the communications MFE which has _no translations_ at the moment so there's no step backward.

Once this PR is merged and used, we'll work on supporting the `i18n-merge.js` if needed. Our understanding that Atlas provides a superior workflow which makes `i18n-merge.js` not needed -- at least as we understand it.

This has been confirmed with @arbrandes -- tagging him just in case I missed anything.

Is what I'm mentioning a blocker to merging this PR?

> Also, this change is going to break the build for all users who do not run your fork of the communications app, right? We should definitely avoid this, one way or another.

No, of course not. The fork is being used merely for reviewing this pull request, we are fixing this on upstream and it's pending merge: 

 - Palm: https://github.com/openedx/frontend-app-communications/pull/156
 - Quince: Pull request will be created.

Comment (OmarIthawi) [reply tutormfe/templates/mfe/build/mfe/Dockerfile:75]: @regisb @arbrandes The https://github.com/openedx/frontend-app-communications/pull/156 pull request has been merged into `palm.master`. We need the following updates to make Tutor MFE support Atlas without needing a fork:

 - Tag `open-release/palm.4` on the communications MFE.
 - Update Tutor open edx version `open-release/palm.2` --> `open-release/palm.4`

This is needed to fix the i18n for the communications app on Palm.

Once this is merged, it will also cause Quince to be supported. The good news that Atlas support already exists in the quince branch: https://github.com/openedx/frontend-app-communications/blob/3e7ea6a3e5d5c35bf508fc7fab067fc5be836e75/Makefile#L52-L62

Comment (regisb) [reply tutormfe/templates/mfe/build/mfe/Dockerfile:75]: > This will affect the communications MFE which has no translations at the moment so there's no step backward.

Well, it will be a breaking change for users who implemented their own translation strings for the communications MFE. It's a niche use case though, so maybe we can just ignore it.

I agree that Atlas should be a much better solution than that weird i18s-merge.js script. Still, we need to provide users a way to customise translation strings. How are we going to do that in the tutor-mfe plugin?

> Tag open-release/palm.4 on the communications MFE.

Publishing a new release tag is a big deal. Do you really need Atlas in Palm? If yes we can ask the BTR to publish a new tag. But maybe it would be sufficient to support Atlas in Quince? You tell me.

Comment (OmarIthawi) [reply tutormfe/templates/mfe/build/mfe/Dockerfile:75]: > Well, it will be a breaking change for users who implemented their own translation strings for the communications MFE. It's a niche use case though, so maybe we can just ignore it.
> I agree that Atlas should be a much better solution than that weird i18s-merge.js script. Still, we need to provide users a way to customise translation strings. How are we going to do that in the tutor-mfe plugin?

For the communications MFE i18n-merge.js is mostly unused due to completely missing translations. Based on the feedback we've had, no one translated it except for Pierre Mailhot which doesn't use the i18n-merge. For other MFEs, yes it makes sense to add this support.

We've discussed this issue in call with Adolfo and we decided that this iteration won't support i18n merge.

Please let me know if that's still a blocker.

> Publishing a new release tag is a big deal. Do you really need Atlas in Palm? If yes we can ask the BTR to publish a new tag. But maybe it would be sufficient to support Atlas in Quince? You tell me.

It's a big deal indeed. Let's refocus our efforts into Quince. I'll check with @sambapete on the idea of skipping Palm for Tutor MFE support, which would leave the Palm users no way to support Atlas except by cherry-picking this pull request. Which isn't pretty bad as long as we land it into Quince.

Comment (OmarIthawi) [reply tutormfe/templates/mfe/build/mfe/Dockerfile:75]: Thanks @sambapete, Quince it is! We're close to do so -- hopefully :)

**Comment by OmarIthawi** [comment]
Status: This is pending the Quince upgrade https://github.com/overhangio/tutor-mfe/pull/156 so I'm presuming that this PR is blocked until the following PR is merged:

 - https://github.com/overhangio/tutor-mfe/pull/156

cc: @regisb @sambapete @arbrandes

**Comment by regisb** [comment]
Any change that would added to nightly now would also go to Quince. Do you want to merge this now in nightly?

File: tutormfe/templates/mfe/build/mfe/Dockerfile
Comment (arbrandes) [reply tutormfe/templates/mfe/build/mfe/Dockerfile:75]: +1 for Quince.

We can still do Palm.4, but it should come _after_ getting Quince up and running.

**Comment by arbrandes** [comment]
+1 for getting this into nightly.

**Comment by OmarIthawi** [comment]
Nightly it is! I'll merge and get back to you!

**Comment by OmarIthawi** [comment]
@regisb @arbrandes I've tested this plugin and it worked well without plugins or settings.

<kbd>image</kbd>

The only remaining pull request has been merged into `master`:

 - https://github.com/openedx/frontend-app-communications/pull/164#issuecomment-1813191272

Please let me know if there's anything blocking this PR from getting merged.

**Comment by OmarIthawi** [comment]
@regisb @arbrandes @brian-smith-tcril I've tested this plugin and it worked well without plugins or settings.

<kbd>image</kbd>

The only remaining pull request has been merged into `master`:

 - https://github.com/openedx/frontend-app-communications/pull/164#issuecomment-1813191272

Please let me know if there's anything blocking this PR from getting merged.

**Review by arbrandes — APPROVED** [review]
Works for me!

**Comment by arbrandes** [comment]
@OmarIthawi, we're missing `openedx-atlas` in frontend-app-communications as it stands in Quince.  It was introduced in `master` via https://github.com/openedx/frontend-app-communications/pull/164.  I created a backport: https://github.com/openedx/frontend-app-communications/pull/176.  Objections?

**Pull request merged**
